petastorm/spark/spark_dataset_converter.py

```diff
@@ -0,0 +1,289 @@
+# limitations under the License.
+
+import atexit
+import os
+import shutil
+import threading
+import uuid
+import warnings
+
+from pyarrow import LocalFileSystem
+from pyspark.sql.session import SparkSession
+from six.moves.urllib.parse import urlparse
+
+from petastorm import make_batch_reader
+from petastorm.fs_utils import FilesystemResolver
+
+DEFAULT_ROW_GROUP_SIZE_BYTES = 32 * 1024 * 1024
+
+
+def _get_spark_session():
+    return SparkSession.builder.getOrCreate()
+
+
+def _delete_cache_data(dataset_url):
+    """
+    Delete the cache data in the underlying file system.
+    """
+    resolver = FilesystemResolver(dataset_url)
+    fs = resolver.filesystem()
+    parsed = urlparse(dataset_url)
+    if isinstance(fs, LocalFileSystem):
+        # pyarrow has a bug: LocalFileSystem.delete() is not implemented.
+        # https://issues.apache.org/jira/browse/ARROW-7953
+        # We can remove this branch once ARROW-7953 is fixed.
+        local_path = parsed.path
+        if os.path.exists(local_path):
+            shutil.rmtree(local_path, ignore_errors=False)
+    else:
+        if fs.exists(dataset_url):
+            fs.delete(dataset_url, recursive=True)
+
+
+def _delete_cache_data_atexit(dataset_url):
+    try:
+        _delete_cache_data(dataset_url)
+    except BaseException:
+        warnings.warn('delete cache data {url} failed.'.format(url=dataset_url))
+
+
+class SparkDatasetConverter(object):
+    """
+    A `SparkDatasetConverter` object holds one materialized spark dataframe and
+    can be used to make one or more tensorflow datasets or torch dataloaders.
+    The `SparkDatasetConverter` object is picklable and can be used in remote
+    processes.
+    See `make_spark_converter`
+    """
+
+    def __init__(self, cache_dir_url, dataset_size):
+        """
+        :param cache_dir_url: A string denoting the path to store the cache
+            files.
+        :param dataset_size: An int denoting the number of rows in the
+            dataframe.
+        """
+        self.cache_dir_url = cache_dir_url
+        self.dataset_size = dataset_size
+
+    def __len__(self):
+        """
+        :return: dataset size
+        """
+        return self.dataset_size
+
+    def make_tf_dataset(self):
+        """
+        Make a tensorflow dataset.
+
+        This method will do the following two steps:
+          1) Open a petastorm reader on the materialized dataset dir.
+          2) Create a tensorflow dataset based on the reader created in (1)
+
+        :return: a context manager for a `tf.data.Dataset` object.
+                 when exit the returned context manager, the reader
+                 will be closed.
+        """
+        return _tf_dataset_context_manager(self.cache_dir_url)
+
+    def delete(self):
+        """
+        Delete cache files at self.cache_dir_url.
+        """
+        _delete_cache_data(self.cache_dir_url)
+
+
+class _tf_dataset_context_manager(object):
```

Comment:
In Python you'd typically use camel case for class names (unless this is something already done elsewhere in Petastorm).